File: App/Resources/Algorithms/introsort.js
// Substituted for the originally-planned TwinSort (hybrid/medium), whose real-world implementation
// (initial "twin swap" pre-pass + bottom-up tail-merging with a separate half-size swap buffer)
// proved far more intricate than its template length suggested. IntroSort reuses building blocks
// already validated elsewhere in this batch: ternary-quicksort-style strict comparisons, 0-indexed
// heapsort (maxheapsort.js), and swap-based insertion sort (insertionsort.js).
//
// partition() uses raw engine.getValue() comparisons against a copied pivot scalar (no recorded
// marks) rather than engine.compare — the pivot value is a plain lifted scalar here, not a live
// stable index reference (unlike the ternary quicksorts' pivot-at-hi/hi-1 trick), so it falls under
// the same "held value" tradeoff as insertion/shell sort. This matches ArrayV itself, which uses
// Reads.compareValues (not compareIndices) for the equivalent checks.
function sort(engine) {
    const n = engine.count();
    const sizeThreshold = 16;

    function medianOf3(left, mid, right) {
        if (!engine.compare(right, left)) {
            engine.swap(left, right);
        }
        if (!engine.compare(mid, left)) {
            engine.swap(mid, left);
        }
        if (!engine.compare(right, mid)) {
            engine.swap(right, mid);
        }
        return mid;
    }

    function partition(lo, hi, pivotValue) {
        let i = lo, j = hi;
        while (true) {
            while (engine.getValue(i) < pivotValue) i++;
            j--;
            while (pivotValue < engine.getValue(j)) j--;
            if (!(i < j)) return i;
            engine.swap(i, j);
            i++;
        }
    }

    function heapSortRange(lo, hi) {
        const size = hi - lo;
        function siftDown(root, rangeSize) {
            while (true) {
                let largest = root;
                const left = 2 * root + 1;
                const right = 2 * root + 2;
                if (left < rangeSize && !engine.compare(lo + largest, lo + left)) largest = left;
                if (right < rangeSize && !engine.compare(lo + largest, lo + right)) largest = right;
                if (largest === root) break;
                engine.swap(lo + root, lo + largest);
                root = largest;
            }
        }
        for (let i = Math.floor(size / 2) - 1; i >= 0; i--) siftDown(i, size);
        for (let end = size - 1; end > 0; end--) {
            engine.swap(lo, lo + end);
            siftDown(0, end);
        }
    }

    function introsortLoop(lo, hi, depthLimit) {
        while (hi - lo > sizeThreshold) {
            if (depthLimit === 0) {
                heapSortRange(lo, hi);
                return;
            }
            depthLimit--;
            const mid = lo + Math.floor((hi - lo) / 2);
            const pivotIndex = medianOf3(lo, mid, hi - 1);
            const pivotValue = engine.getValue(pivotIndex);
            const p = partition(lo, hi, pivotValue);
            introsortLoop(p, hi, depthLimit);
            hi = p;
        }
    }

    function insertionSort(start, end) {
        for (let i = start + 1; i < end; i++) {
            let j = i;
            while (j > start && !engine.compare(j, j - 1)) {
                engine.swap(j - 1, j);
                j--;
            }
        }
    }

    function floorLog2(a) {
        return Math.floor(Math.log(a) / Math.log(2));
    }

    introsortLoop(0, n, 2 * floorLog2(n));
    insertionSort(0, n);
}

File: Tools/AlgorithmContentPipeline/introsort.bundle/introsort.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

int medianOf3(int arr[], int left, int mid, int right) {
  if (!(arr[left] >= arr[right])) {
    swap(&arr[left], &arr[right]);
  }
  if (!(arr[left] >= arr[mid])) {
    swap(&arr[left], &arr[mid]);
  }
  if (!(arr[mid] >= arr[right])) {
    swap(&arr[mid], &arr[right]);
  }
  return mid;
}

int partition(int arr[], int lo, int hi, int pivotValue) {
  int i = lo, j = hi;
  while (1) {
    while (arr[i] < pivotValue) i++;
    j--;
    while (pivotValue < arr[j]) j--;
    if (!(i < j)) return i;
    swap(&arr[i], &arr[j]);
    i++;
  }
}

void siftDown(int arr[], int lo, int root, int rangeSize) {
  while (1) {
    int largest = root;
    int left = 2 * root + 1;
    int right = 2 * root + 2;
    if (left < rangeSize && arr[lo + largest] < arr[lo + left]) largest = left;
    if (right < rangeSize && arr[lo + largest] < arr[lo + right]) largest = right;
    if (largest == root) break;
    swap(&arr[lo + root], &arr[lo + largest]);
    root = largest;
  }
}

void heapSortRange(int arr[], int lo, int hi) {
  int size = hi - lo;
  for (int i = size / 2 - 1; i >= 0; i--) siftDown(arr, lo, i, size);
  for (int end = size - 1; end > 0; end--) {
    swap(&arr[lo], &arr[lo + end]);
    siftDown(arr, lo, 0, end);
  }
}

void insertionSort(int arr[], int start, int end) {
  for (int i = start + 1; i < end; i++) {
    int j = i;
    while (j > start && arr[j] < arr[j - 1]) {
      swap(&arr[j - 1], &arr[j]);
      j--;
    }
  }
}

int floorLog2(int a) {
  return (int)floor(log((double)a) / log(2.0));
}

void introsortLoop(int arr[], int lo, int hi, int depthLimit) {
  while (hi - lo > 16) {
    if (depthLimit == 0) {
      heapSortRange(arr, lo, hi);
      return;
    }
    depthLimit--;
    int mid = lo + (hi - lo) / 2;
    int pivotIndex = medianOf3(arr, lo, mid, hi - 1);
    int pivotValue = arr[pivotIndex];
    int p = partition(arr, lo, hi, pivotValue);
    introsortLoop(arr, p, hi, depthLimit);
    hi = p;
  }
}

void sort(int arr[], int n) {
  introsortLoop(arr, 0, n, 2 * floorLog2(n));
  insertionSort(arr, 0, n);
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
